workspaces/api-gateway/src/__tests__/api/products/products.controller.spec.ts

```diff
@@ -0,0 +1,194 @@
+import { categoriesSample } from '__tests__/seeds/categories';
+import { CheckCategoryExistedService } from 'api/categories/services/CheckCategoryExistedService';
+import { CreateProductService } from 'api/products/services/CreateProductService';
+import { CreateCategoryService } from 'api/categories/services/CreateCategoryService';
+import { GetCategoryByKeyService } from 'api/categories/services/GetCategoryByKeyService';
+import { GetRelatedCategoriesService } from 'api/categories/services/GetRelativeCategoryService';
+import { ProductsPagyMetadataVM, ProductVM } from 'api/products/product.vm';
+import type { CreateProductDto } from 'api/products/products.dto';
+import type { CreateCategoryDto } from 'api/categories/categories.dto';
+
+describe('ProductsController (e2e)', () => {
+  let app: INestApplication;
+  let createCategoryService: CreateCategoryService;
+  let createProductService: CreateProductService;
+  let dataSource: DataSource;
+  let mapper: Mapper;
+  let product: Product;
+  let category: Category;
+
+  const createProductDto: CreateProductDto = {
+    categoryId: null,
+    name: productSample.name,
+  };
+
+  const createCategoryDto: CreateCategoryDto = {
+    name: categoriesSample.name,
+    parentId: null,
+  };
+
+  beforeAll(async () => {
+    const { module, app: appInstance } = await createTestingModule({
+      imports: [ProductsModule],
+      entities: [Category, Product, ProductsCategories],
+      providers: [
+        CreateCategoryService,
+        CheckCategoryExistedService,
+        GetRelatedCategoriesService,
+        GetCategoryByKeyService,
+      ],
+    });
+
+    dataSource = module.get(DataSource);
+    createCategoryService = module.get(CreateCategoryService);
+    createProductService = module.get(CreateProductService);
+
+    mapper = module.get<Mapper>('automapper:nestjs:default');
+
+    app = appInstance;
+    app.setGlobalPrefix('api');
+    await app.init();
+  });
+
+  beforeEach(async () => {
+    category = await createCategoryService.exec(createCategoryDto);
+    createProductDto.categoryId = category.id;
+    product = await createProductService.exec(createProductDto);
+  });
+
+  afterEach(async () => {
+    const QueryRunner = dataSource.createQueryRunner();
+
+    await QueryRunner.manager.query('TRUNCATE products_categories CASCADE');
+    await QueryRunner.manager.query('TRUNCATE products CASCADE');
+    await QueryRunner.manager.query('TRUNCATE categories_closure CASCADE');
+    await QueryRunner.manager.query('TRUNCATE categories CASCADE');
+  });
+
+  afterAll(async () => {
+    await dataSource.destroy();
+    await app.close();
+  });
+
+  describe('POST /api/products', () => {
+    describe('with valid product name', () => {
+      it('returns product', async () => {
+        const resultMapping = await mapper.mapAsync(
+          product,
+          Product,
+          ProductVM,
+        );
+
+        jest.spyOn(Date.prototype, 'getTime').mockImplementation(() => 123);
```

Comment:
this mock is used on multiple places, should include this one inside `createModule` helper